internal/expr/query.go

```diff
@@ -36,92 +36,120 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
+			for _, typeElement := range pe.inputs {
+				if t.Name() == typeElement.outerType().Name() {
+					return nil, fmt.Errorf("type %s not found, have %s", typeElement.outerType().String(), t.String())
 				}
 			}
-
 			return nil, fmt.Errorf("%s not referenced in query", t.Name())
 		}
-
 	}
 
 	// Query parameteres.
 	qargs := []any{}
-
-	for i, in := range pe.inputs {
-		v, ok := typeValue[in.structType]
+	for i, typeElement := range pe.inputs {
+		v, ok := typeValue[typeElement.outerType()]
 		if !ok {
-			return nil, fmt.Errorf(`type %s not found, have: %s`, in.structType.Name(), strings.Join(typeNames, ", "))
+			return nil, fmt.Errorf(`type %s not found, have: %s`, typeElement.outerType().Name(), strings.Join(typeNames, ", "))
+		}
+		switch te := typeElement.(type) {
+		case field:
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(te.index).Interface()))
+		case mapKey:
+			// MapIndex returns a zero value of the key is not in the map so we
+			// need to check the MapKeys().
+			var val reflect.Value
+			for _, key := range v.MapKeys() {
+				if key.String() == te.name {
+					val = v.MapIndex(reflect.ValueOf(te.name))
+					break
+				}
+			}
+			if !val.IsValid() {
+				return nil, fmt.Errorf(`map does not contain key %q`, te.name)
+			}
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(i), val.Interface()))
 		}
-		named := sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(in.index).Interface())
-		qargs = append(qargs, named)
 	}
-
 	return &QueryExpr{outputs: pe.outputs, sql: pe.sql, args: qargs}, nil
 }
 
+type MapDecodeInfo struct {
+	m         reflect.Value
+	valuePtrs []any
+	keyIndex  map[string]int
+}
+
 // ScanArgs returns list of pointers to the struct fields that are listed in qe.outputs.
 // All the structs mentioned in the query must be in outputArgs.
 // All outputArgs must be structs.
-func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, error) {
+func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, []*MapDecodeInfo, error) {
+	var typesInQuery = []string{}
+	var inQuery = make(map[reflect.Type]bool)
+	for _, typeElement := range qe.outputs {
+		if ok := inQuery[typeElement.outerType()]; !ok {
+			inQuery[typeElement.outerType()] = true
+			typesInQuery = append(typesInQuery, typeElement.outerType().Name())
+		}
+	}
+
+	var mapDecodeInfos = make(map[reflect.Type]*MapDecodeInfo)
+	var typeDest = make(map[reflect.Type]reflect.Value)
 	outputVals := []reflect.Value{}
 	for _, outputArg := range outputArgs {
 		if outputArg == nil {
-			return nil, fmt.Errorf("need pointer to struct, got nil")
+			return nil, nil, fmt.Errorf("need map or pointer to struct, got nil")
 		}
 		outputVal := reflect.ValueOf(outputArg)
-		if outputVal.Kind() != reflect.Pointer {
-			return nil, fmt.Errorf("need pointer to struct, got %s", outputVal.Kind())
+		k := outputVal.Kind()
```

Comment:
Would it be a good idea to have a switch block for Map and Pointer types?

```Go
switch k := outputVal.Kind() {
case reflect.Pointer:
    // do Pointer-specific error stuff
    // then do Pointer stuff
case reflect.Map:
    // do Map-specific error stuff
    // then do Map stuff
default:
    // do other error stuff
}
```

Reply:
I was thinking of doing this, but we have the outputVal.Elem() in the middle that changes the Kind of outputVal half way though and makes it all a bit messy.